[#4444] Add coverator for publishing coverage reports.

bgola opened the pull request:
Scope
=====

Update brink to use coverator when publishing coverage reports.


Changes
=======

I've updated the coverage_publish method.

I've also added a configuration to SETUP to set the address where to upload reports.

Also fixed the revision property in BrinkGit which was returning the tree, not the revision hash.


How to try and test the changes
===============================

reviewers: @adiroiban 

Please see if changes make sense.

adiroiban commented on pavement.py:
```diff
@@ -124,8 +124,9 @@
-    'coverage==4.0.3',
+    'coverage==4.4.1',
     'codecov==2.0.3',
+    'chevah-coverage==0.1.bgola6',
```
this will need to have a final version... like 0.1.0.

adiroiban commented on brink/pavement_commons.py:
```diff
@@ -456,38 +457,31 @@ def coverage_publish():
     """
-    from pkg_resources import load_entry_point
-    import coverage
+    from chevah.coverage.client import upload_coverage
```
this is fine, but I would stil like to keep the old code as `codecov_publish` since in some projects (especially the public one like compat and keycert) we would still publish to codecov.io

codecov.io was working for compat, it only had issues with chevah/server

adiroiban commented on brink/configuration.py:
```diff
@@ -64,6 +64,8 @@
         # List of nose arguments passed to all tests.
         'nose_options': [],
+        # URL for publishing the coverage reports.
+        'coverate_url': 'http://172.20.245.1:8080'
```
maybe in brink we should leave this as None and have each project configure the url.... or put here the DSN value like coverage.chevah.com :)

bgola commented:
@adiroiban tests are failing because there is no http://coverage.chevah.com/

but in the case of brink we might want to use codecov still, right?

needs-review

adiroiban commented:
Well, we can always update the SETUP value in pavement.py in brink.

In brink I am expecting to see the coverage sent to both codecov.io and to our coverage tool so that we can test.

Please update the branch so that we get a functional coverage report here. Thanks!

needs-changes

bgola commented:
I think it is now sending to both codecov and our coverage. I don't know why codecov is not reporting to github tho.

needs-review

adiroiban commented:
codecov is buggy, or we don't know how to use it. This is why I wanted our own coverage.

I have observed that reports are sent only when all the other tests are green.

adiroiban commented on brink/pavement_commons.py:
```diff
@@ -324,6 +325,8 @@ def test_os_dependent(args):
     if codecov_token:
         # Only publish if we have a token.
+        call_task('codecov_publish')
+    if SETUP['test']['coverate_url']:
```
this should be remove, together with previous publishing code.

We no longer implicitly publish the coverage.

Rather, the coverage is explicitly published in the test_ci of each project

adiroiban commented on brink/pavement_commons.py:
```diff
@@ -490,6 +493,38 @@ def coverage_publish():
+@task
+def coverage_publish():
+    """
+    Send the coverage report.
```
We need a name :)
Send the coverage report to OUR_SERVICE_NAME.

bgola commented:
i have used `coverator` as the name for the new tool. 

updated the code here to use the new name, and updated pavement.py to explicitly publish the coverage data.

now, if the plan is to keep using codecov, maybe we should keep coverage_publish for codecov and renew the new one as coverator_publish, in this way we only update the projects we want to use coverator instead of codecov?

needs-review

bgola commented:
ah, i'm trying to understand why the `paver deps` is failing on some systems.

bgola commented:
@adiroiban there were three different failures, one of them was that it was trying to publish reports on systems that do not run the tests with coverage. I've fixed this by disabling coverage for py3 and os-independent.

now, there is this `unicode` problem on the windows machines, and I think you have fixed this on compat?

the other error is that it fails on some systems to get the `git origin`. I did ssh into the servers and yes, some builders .git/config doesn't contain the remotes/origin entry. do you know why some have and others dont?

thanks!

bgola commented:
all tests are passing now. 

this error occurs when I upgrade python to 2.7.14 on sles-10: 

https://chevah.com/buildbot/builders/brink-sles-10/builds/23/steps/test/logs/stdio

so I kept it on 2.7.13.

needs-review

adiroiban commented:
I am using this as a review for the coverator tools.

coverage/project should be coverator/project.

For coverage/project  we need to fail only if coverage is decreasing as compared with master.

Now that we see that coverage/project/diff fails, please write a simple test for the change, so see that diff is observed.

-------

Since this is the brink repo, we should also upload to codecov.io to check that the codecov code works.

You can send to codecov.io only on windows and linux... on aix / hpux / solaris might fail due to missing encryption

adiroiban commented on pavement.py:
```diff
@@ -322,21 +389,34 @@ def test_ci(args):
+        # Only publish to codecov if we have a token.
+        if pave.os_name in ['linux', 'ubuntu1604', 'archlinux', 'windows']:
+            call_task('codecov_publish')
+        if SETUP['test']['coverator_url']:
```
no need for this condition, as this is always set in this pavement.py file

adiroiban commented:
compat is already on 2.7.14 so I think that it should also work for brink

bgola commented:
@adiroiban it is working now as we discussed in the meeting, coverator sends the xml to codecov and codecov will send the notification to github even if there is a test failing.

bgola commented:
also, coverator is not sending the github notification anymore.

i will do the same on compat and if it works maybe i should open a PR on server and do it as well?

adiroiban commented:
you forgot to add coverator==0.1.0 dependecy in setup.py for brink :) this is why it can't find coverator on some builders

also, brink should use the entry point instead of importing it directly

adiroiban commented:
coverator publish needs a bit more status... like the address where it push and which env vars were used... and if the upload was successful or not.

I have this PR for server https://github.com/chevah/server/pull/1280 and I only 3 uploads http://172.20.245.1:8080/chevah/server/pr/1280

adiroiban commented:
I saw where the missing PR were gone.

Note that all reports should be done based on SHA, and this should be a full one.
Don't trust the clients for generation  a short version as you can get differrent results

see http://172.20.245.1:8080/chevah/server/commit/

The reports should be generated per sha, so the PR or a branch should not have a separate reports, but just symlinks.

For GitHub status, SHA commit is all that matter.

PR and branch names are only used for easy navigation for serving the HTML files.

At http://172.20.245.1:8080/chevah/server/pr/1280/ I was expecting to also see the HTML files for coverage, so that I could check and browse the coverage

bgola commented:
> you forgot to add coverator==0.1.0 dependecy in setup.py for brink :) this is why it can't find coverator on some builders

but there are no packages in setup.py, they are all in pavement.py. i tried to do exactly what was done  with the codecov package.

also I'm still not sure about this:

> 19:58 <bgola> one thing im a bit confused by: pavement.py has TEST and RUN and BUILD PACKAGES sections
> 19:58 <bgola> brink includes coverator in the TEST section
> 19:58 <bgola> but tecnically it is needed when using brink, by compat or server
> 19:58 <bgola> so coverator should be in the RUN section?


and what is the reason for using `load_entry_point` instead of importing `coverator` directly?

adiroiban commented on brink/pavement_commons.py:
```diff
@@ -490,6 +485,39 @@ def coverage_publish():
+    repository = re.split(r'github.com[/:]', SETUP['repository']['github'])[1]
+
+    builder_name = os.environ.get('BUILDER_NAME', pave.getHostname())
+    github_pull_id = os.environ.get('GITHUB_PULL_ID', '')
```
in qm.py we have a helper like

```
def _get_environment(name, default=_REQUIRED):
    """
    Get environment variable.
    """
    value = os.environ.get(name, default)
    if value is _REQUIRED:
        raise AssertionError(
            'Variable %s not found in environment !' % (name))
    return value
```

maybe it should be refactored into a public code and use it in every part where we read some env var which is required

adiroiban commented:
Something is strange.

Report for `codecov/project/tests` is 100% but here I see 85% https://codecov.io/gh/chevah/brink/pull/113/tree/brink/tests

adiroiban commented:
I have enabled it for chevah/server but is not working https://github.com/chevah/server/pull/1313
I see more than 10 reports here http://172.20.245.1:8080/chevah/server/pr/1313/

The problem seems to be due to the fact that  we don't have a report for master https://codecov.io/gh/chevah/server/pull/1313...I will try to generate one.

I tried to download the report files and to use them locally... but I don't know why they produce no reports

bgola commented:
I was checking the coverator-server, problem is that chevah/server is a private repository and we were working only with public repos so far (compat and brink), so it fails to clone the server repository to generate the report.

adiroiban commented:
> I was checking the coverator-server, problem is that chevah/server is a private repository and we were working only with public repos so far (compat and brink), so it fails to clone the server repository to generate the report.

you should use the same token as the one used to push the status. if the token is generated by your account then it has permissions to clone the server repo

bgola commented:
i have fixed the coverator server, it should work with chevah/server now. 

https://github.com/chevah/server/pull/1313

adiroiban commented:
something is not ok with the coverage upload task. see these results
 https://chevah.com/buildbot/builders/server-gk-review/builds/127

coverage upload should have a timeout of maximum 30 seconds, and close if server is not reachable.
and if coverage upload fails, the tests should not fail (at leas for now, while we are still testing coverage reporting) 

:)

bgola commented:
I guess this is also good for review.

I have publish the new brink 0.67.3, as soon as this is approved we can update the version for compat and server so it uses the new coverator with timeout support.

need-review

adiroiban commented on brink/pavement_commons.py:
```diff
@@ -515,7 +515,10 @@ def coverator_publish():
+        try:
+            upload_coverage(*args, url=SETUP['test']['coverator_url'])
+        except Exception:
+            print('Failed to upload coverage data:', sys.exc_info[0])
```
why not use

```
except Exception as error:
    print('bla %s' % (error,)
```
no need to access the sys.exc_info

and this assumes that upload_coverage in 0.1.1 will return with a timeout after 30 seconds

as otherwise, it it takes more than 60 seconds to return, buidbot will kill it with an error.
